fades/cache.py

```diff
@@ -102,5 +99,42 @@ def store(self, installed_stuff, metadata, interpreter):
         }
         logger.debug("Storing installed=%s metadata=%s interpreter=%s",
                      installed_stuff, metadata, interpreter)
-        with open(self.filepath, 'at', encoding='utf8') as fh:
-            fh.write(json.dumps(new_content) + '\n')
+        with self.lock_cache():
+            self._write_cache([json.dumps(new_content)], append=True)
+
+    def remove(self, env_path):
+        """Remove metadata for a given virtualenv from cache."""
+        with self.lock_cache():
+            cache = self._read_cache()
+            logger.debug("Removing virtualenv from cache: %s" % env_path)
+            lines = [
+                line for line in cache
+                if json.loads(line).get('metadata', {}).get('env_path') != env_path
+            ]
+            self._write_cache(lines)
+
+    @contextmanager
+    def lock_cache(self):
+        lock_file = self.filepath + '.lock'
+        with open(lock_file, 'a') as fh:
+            fcntl.flock(fh, fcntl.LOCK_EX)
+            yield
+            fcntl.flock(fh, fcntl.LOCK_UN)
+        if os.path.exists(lock_file):
+            os.remove(lock_file)
+
+    def _read_cache(self):
+        """Read virtualenv metadata from cache."""
+        if os.path.exists(self.filepath):
+            with open(self.filepath, 'rt', encoding='utf8') as fh:
+                lines = [x.strip() for x in fh]
+        else:
+            logger.debug("Index not found, starting empty")
+            lines = []
+        return lines
+
+    def _write_cache(self, lines, append=False):
```

Comment:
but we never call it with append in False, or not adding the append parameter, right?